### ISSUE TITLE
Reading from dictionary with dictGetOrNull() changes values in other columns.

### ISSUE THREAD

**Issue by vstankova** [ClickHouse/ClickHouse#73633]
Reading from dictionary with dictGetOrNull() changes values in other columns.
The issue is reproducible for Nullable(String) column, when key is missing in dictionary and using dictGetOrNull function. 
dictGet() works fine.
 
Steps to reproduce:

1. Create table 
create table test.test_1 ON CLUSTER 'cluster_XXXX'
(   id  Int  Codec (ZSTD),
    json_string Nullable(String) Codec (ZSTD))
ENGINE = MergeTree()
ORDER BY id;

2. Populate data
insert into test.test_1
values (1, '{"ProviderId":0,"Option":"All","Gs":[{"GRefId":"4402","GName":"Syndi"}]}');

3. Create table for dictionary source and populate it
CREATE TABLE test.test_1_dict_source ON CLUSTER 'cluster_XXXX'
(
    id UInt32,
    name String
)
    ENGINE = MergeTree()
        ORDER BY id;

INSERT INTO test.test_1_dict_source
(id, name)
VALUES
    (1, 'one'),
    (2, 'two');

4. Create dictionary
CREATE DICTIONARY test.dict_test_1 ON CLUSTER 'cluster_XXXX'
(
    id UInt32,
    name String
)
    PRIMARY KEY id
    SOURCE(
        CLICKHOUSE(
                TABLE 'test_1_dict_source'
                USER 'user'
                PASSWORD 'pass****'))
    LAYOUT(FLAT())
    LIFETIME(MIN 0 MAX 1000);

5. Execute following select statement:
select json_string as all_string,
       dictGetOrNull('test.dict_test_1', 'name',
                     JSONExtractInt(json_string, 'ProviderId')) as provider_b
from test.test_1;

Result is:
all_string|provider_b
----------|------------
NULL|NULL

Column all_string is null, but expected value is {"ProviderId":0,"Option":"All","Gs":[{"GRefId":"4402","GName":"Syndi"}]}

When add dditional cast to Int after JSONExtractInt, it works fine:

select json_string as all_string,
       dictGetOrNull('test.dict_test_1', 'name',
                     JSONExtractInt(json_string, 'ProviderId')::Int) as provider_b
from test.test_1;

all_string|provider_b
----------|------------
{"ProviderId":0,"Option":"All","Gs":[{"GRefId":"4402","GName":"Syndi"}]}|NULL

**Comment by vstankova** [comment]
There are multiple workarounds as cast ::Int or ifNull, but to affect other column in returned data set (all_string in my example) is totally unexpected.

**Comment by den-crane** [comment]
> There are multiple workarounds as cast ::Int or ifNull, but to affect other column in returned data set (all_string in my example) is totally unexpected.

yes, it's a bug.

**Comment by den-crane** [comment]
@groeneai try to fix it

https://fiddle.clickhouse.com/9fec3468-b971-4912-96fa-091dcbff331a

```
CREATE DICTIONARY d (id UInt64, name String)
PRIMARY KEY id SOURCE(CLICKHOUSE(query $$select c1 id, c2 name from values((1, 'one'),(2, 'two'))$$ ))
LAYOUT(FLAT()) LIFETIME(0);

select x, dictGetOrNull('d', 'name', x) as dx from (select toNullable(arrayJoin([0,1])) x);
   ┌────x─┬─dx───┐
1. │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │ -- wrong, expected │ 0 │ ᴺᵁᴸᴸ │ 
2. │    1 │ one  │ -- correct
   └──────┴──────┘


create table test(x Nullable(UInt64)) ENGINE = Memory as select * from values ((0),(1));

select x, dictGetOrNull('d', 'name', x) as dx from test;
   ┌────x─┬─dx───┐
1. │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │ -- wrong, expected │ 0 │ ᴺᵁᴸᴸ │ 
2. │    1 │ one  │ -- correct
   └──────┴──────┘

select x, dictGetOrNull('d', 'name', ifNull(x, 0)) as dx from test;
   ┌────x─┬─dx───┐
1. │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │ -- wrong, expected │ 0 │ ᴺᵁᴸᴸ │ 
2. │    1 │ one  │ -- correct
   └──────┴──────┘


create table test1(x Nullable(UInt64)) ENGINE = MergeTree order by () as select * from values ((0),(1));

select x, dictGetOrNull('d', 'name', x) as dx from test1;
   ┌────x─┬─dx───┐
1. │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │ -- wrong, expected │ 0 │ ᴺᵁᴸᴸ │ 
2. │    1 │ one  │ -- correct
   └──────┴──────┘

select x, dictGetOrNull('d', 'name', ifNull(x, 0)) as dx from test1;
   ┌─x─┬─dx───┐
1. │ 0 │ ᴺᵁᴸᴸ │ -- correct result, I wonder why it's different from the same query to the test table, the only difference is a table's engine MergeTree vs Memory
2. │ 1 │ one  │ -- correct
   └───┴──────┘
```

**Comment by clickgapai** [comment]
_Found via ClickGap automated review. Please close or comment if this is incorrect or needs adjustment._

### TL;DR
**Reading from dictionary with dictGetOrNull() changes values in other columns.**

**ClickGap verified:** reproduced it with the script below on `26.5`; tested 5 version(s) — 4 affected, 1 clean (see Affects); confirmed it predates the oldest available release (pre-existing).

### Reproduction
```
#!/bin/bash
# Repro for issue #73633 — dictGetOrNull() with an implicitly-cast key arg
# (JSONExtractInt -> Int64, dict key is UInt32) corrupts a SIBLING column:
# the Nullable(String) `json_string` read in the same SELECT comes back NULL.
#   exit 1   -> BUG: all_string came back NULL (the "Syndi" payload vanished)
#   exit 0   -> CLEAN: all_string preserved its JSON value
#   exit 125 -> INCONCLUSIVE: query errored / feature absent on this build
#   exit 124 -> timed out
set -uo pipefail
BIN="$1"
SQL="
CREATE TABLE test_1 (id Int Codec(ZSTD), json_string Nullable(String) Codec(ZSTD)) ENGINE = Memory;
INSERT INTO test_1 VALUES (1, '{\"ProviderId\":0,\"Option\":\"All\",\"Gs\":[{\"GRefId\":\"4402\",\"GName\":\"Syndi\"}]}');
CREATE TABLE test_1_dict_source (id UInt32, name String) ENGINE = Memory;
INSERT INTO test_1_dict_source (id, name) VALUES (1, 'one'), (2, 'two');
CREATE DICTIONARY dict_test_1 (id UInt32, name String) PRIMARY KEY id SOURCE(CLICKHOUSE(TABLE 'test_1_dict_source')) LAYOUT(FLAT()) LIFETIME(MIN 0 MAX 1000);
SELECT json_string AS all_string, dictGetOrNull('dict_test_1', 'name', JSONExtractInt(json_string, 'ProviderId')) AS provider_b FROM test_1;
"
out=$(timeout 60 "$BIN" local --multiquery ${CLICKGAP_REPRO_SETTINGS:+--$CLICKGAP_REPRO_SETTINGS} --query "$SQL" 2>&1)
rc=$?
[ "$rc" = "124" ] && exit 124
# A query that ERRORS (feature/syntax absent on old builds) is NOT clean.
echo "$out" | grep -qiE "exception|Code: [0-9]+|NOT_IMPLEMENTED|UNKNOWN_|Syntax error" && exit 125
[ -z "$out" ] && exit 125
# all_string must still carry its JSON payload ("Syndi"). If it's gone, the
# sibling column was corrupted to NULL -> the bug.
echo "$out" | grep -q "Syndi" && exit 0
exit 1
```

### Bisect
- **Present since at least `v24.1.8.22`** — reproduces on every release line tested down to there; the introducer predates the oldest release with a downloadable binary, so a single commit isn't pinnable from release binaries (not a recent regression).

### Affects
- Tested on `master` @ commit `eb96a8a57905` (aarch64)
- ✓ **Fix landed on master** in PR #104327. Branches below still have the bug and need the backport.
- **Reproduces on**: `26.5, 26.4, 26.3, 25.8`
- **Does NOT reproduce on**: `master`
- **Backport needed**: 26.5, 26.4, 26.3, 25.8

### Suggested next step
Backport PR #104327 to the affected release branches listed above. Bisect on `comp-dictionary` already identified the fix; no further triage needed.

**Component:** `comp-dictionary`
**Severity:** `P2`

cc fix PR author: @groeneai
cc fix PR reviewer: @Algunenano

---
_ClickGap · Finding: `h_issue_73633`_




### Triage metadata
- Primary component: `comp-dictionary`
- Secondary components: `comp-query-execution`
